influxdb3_write/src/wal.rs

```diff
@@ -1,44 +1,642 @@
+    #[error("length mismatch for segment {segment_id:?}: expected {expected}, got {actual}")]
+    LengthMismatch {
+        segment_id: SegmentId,
+        expected: u32,
+        actual: u32,
+    },
+
+    #[error("checksum mismatch for segment {segment_id:?}: expected {expected}, got {actual}")]
+    ChecksumMismatch {
+        segment_id: SegmentId,
+        expected: u32,
+        actual: u32,
+    },
+
+    #[error("invalid segment file name {0:?}")]
+    InvalidSegmentFileName(String),
+
+    #[error("unable to parse segment id from file: {source}")]
+    UnableToParseSegmentId {
+        file_name: String,
+        source: std::num::ParseIntError,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
 
 #[derive(Debug)]
 pub struct WalImpl {
-    #[allow(dead_code)]
-    path: PathBuf,
+    root: PathBuf,
 }
 
 impl WalImpl {
-    pub fn new(path: impl Into<PathBuf>) -> Self {
-        Self { path: path.into() }
+    pub fn new(path: impl Into<PathBuf>) -> Result<Self> {
+        let root = path.into();
+        info!(wal_dir=?root, "Ensuring WAL directory exists");
+        std::fs::create_dir_all(&root)?;
+
+        // ensure the directory creation is actually fsync'd so that when we create files there
+        // we don't lose them (see: https://www.usenix.org/system/files/conference/osdi14/osdi14-paper-pillai.pdf)
+        File::open(&root)
+            .expect("should be able to open just-created directory")
+            .sync_all()
+            .expect("fsync failure");
+
+        Ok(Self { root })
+    }
+
+    fn open_segment_writer(&self, segment_id: SegmentId) -> Result<Box<dyn WalSegmentWriter>> {
+        let writer = WalSegmentWriterImpl::new_or_open(self.root.clone(), segment_id)?;
+        Ok(Box::new(writer))
+    }
+
+    fn open_segment_reader(&self, segment_id: SegmentId) -> Result<Box<dyn WalSegmentReader>> {
+        let reader = WalSegmentReaderImpl::new(self.root.clone(), segment_id)?;
+        Ok(Box::new(reader))
+    }
+
+    fn segment_files(&self) -> Result<Vec<SegmentFile>> {
+        let dir = std::fs::read_dir(&self.root)?;
+
+        let mut segment_files = Vec::new();
+
+        for child in dir.into_iter() {
+            let child = child?;
+            let meta = child.metadata()?;
+            if meta.is_file() {
+                let path = child.path();
+                let file_name = path
+                    .file_stem()
+                    .expect("WAL segment files created by InfluxDB3 should have a file stem");
+                let file_name = file_name.to_str().expect("WAL segment files created by InfluxDB3 should have a file stem that is valid UTF-8");
+
+                if let Ok(segment_id) = segment_id_from_file_name(file_name) {
+                    segment_files.push(SegmentFile { segment_id, path });
+                } else {
+                    warn!(
+                        file_name,
+                        "WAL segment file has an invalid file stem, ignoring"
```

Comment:
We crash for the other two situations and this seems like an invariant we would want to uphold. Should we not also crash here?

Reply:
My assumption here is that they have some other file that isn't a segment file in the wal directory, so just ignore it.

Reply:
Probably better not to crash here. There are likely to be troubleshooting situations in prod where it's useful to temporarily rename wal files.

Reply:
Fixed in c47e202d878eb0f6f7f920920627ec052d1c5c7f